cleveref-usedon.dtx

```diff
@@ -817,43 +941,85 @@
 %	Throw an error, if an unrecognised option was used for the
 %	optional argument to this macro.
 %\iffalse
-%%                    % Throw an error, if an unrecognised option was used
-%%                    % for the optional argument to this macro.
+%%            % Throw an error, if an unrecognised option was used
+%%            % for the optional argument to this macro.
 %\fi
 %    \begin{macrocode}
             \msg_new:nnn {cleveref-usedon} { OptionSpellingError }
-                {   \MessageBreak
-                    Spelling~error~\msg_line_context:
-                    \MessageBreak
-                    Did~you~mean~to~pass~option\MessageBreak
-                    'UsedOn'~to~cref~or~Cref?
-                }
+              {
+                \MessageBreak
+                Spelling~error~\msg_line_context:
+                \MessageBreak
+                Did~you~mean~to~pass~option\MessageBreak
+                'UsedOn'~to~cref~or~Cref?
+              }
             \msg_fatal:nn { cleveref-usedon } { OptionSpellingError }
         }
     }
-}%
+  }
 %    \end{macrocode}
 %\end{macro}
 %
 %\begin{macro}{\@@_cref}
 %	This is just a wrapper around \pkg{cleveref}'s |\cref|.
 %	Additionally the \cs{@@_Processor} gets called.
 %    \begin{macrocode}
-\NewDocumentCommand{\@@_cref}{ s o m }{%
-    \IfBooleanTF{#1}{ \@@_origcref*{#3} }{ \@@_origcref{#3} }%
-    \@@_Processor[#2]{#3}
-}%
-%    \end{macrocode}
-%\end{macro}
-%
-%\begin{macro}{\@@_Cref}
-%	This is just a wrapper around \pkg{cleveref}'s |\Cref|.
-%	Additionally the \cs{@@_Processor} gets called.
-%    \begin{macrocode}
-\NewDocumentCommand{\@@_Cref}{ s o m }{%
-    \IfBooleanTF{#1}{ \@@_origCref*{#3} }{ \@@_origCref{#3} }%
-    \@@_Processor[#2]{#3}
-}%
+\bool_new:N \l_@@_already_specified_bool
+\clist_new:N \l_@@_args_clist
+\cs_new:Nn \@@_define_from_orig:NN
+  {
+    \NewDocumentCommand #1 { s o m }
+      {
+        \IfValueTF { ##2 }
+          {
+            \bool_set_false:N \l_@@_already_specified_bool
+            \clist_set:Nn \l_@@_args_clist { ##2 }
+            \clist_put_right:No \l_@@_args_clist { \l_@@_default_tl }
+            \clist_map_inline:Nn \l_@@_args_clist
+              {
+                \str_case:xnT { \str_foldcase:n { ####1 } }
+                  {
+                    {usedon} {}
+                    {uo}     {}
+                    {usedby} {}
+                    {ub}     {}
+                    {usedbyandon} {}
+                    {ubao}        {}
+                  }
+                  {
+                    \clist_remove_all:Nn \l_@@_args_clist { ####1 }
+                    \bool_if:NF \l_@@_already_specified_bool
```

Comment:
If I understand correctly, then code like
```latex
\usepackage[default=UsedOn]{cleveref-usedon}
\cref[UsedBy, UsedByAndOn, ...]{lemma:new_result}
```
(does whitespace after commas matter in a clist?)
would result effectively in
```latex
\cref[UsedBy, ...]{lemma:new_result}
```
As in: The first of possibly multiple `UsedOn`-type local options is applied and overwrites the global default package option whereas further non-`UsedOn`-type options are passed down to the `\origcref` command. Right?

Reply:
Right, it seems so.

If I remember correctly, the reason for me to choose to use the leftmost option instead of rightmost is that once an option is matched, `\__UsedOn_Processor` shall directly print the message, and any further matches would result in printing multiple messages, so simply disabling the rest options seems natural. It is certainly possible to rewrite this part of code to use the rightmost option, I was just following the order given by `\clist_map_inline`. This doesn't matter too much, right? I suppose that there should be at most one `UsedOn`-type option to be specified by users, anyway.